consensus/XDPoS/engines/engine_v2/engine.go

```diff
@@ -319,41 +319,55 @@ func (x *XDPoS_v2) calcDifficulty(chain consensus.ChainReader, parent *types.Hea
 	return big.NewInt(1)
 }
 
-// Copy from v1
+// Check if it's my turm to mine a block. Note: The second return value `preIndex` is useless in V2 engine
 func (x *XDPoS_v2) YourTurn(chain consensus.ChainReader, parent *types.Header, signer common.Address) (int, int, int, bool, error) {
-	snap, err := x.GetSnapshot(chain, parent)
+	x.lock.RLock()
+	defer x.lock.RUnlock()
+
+	round := x.currentRound
+	isEpochSwitch, _, err := x.IsEpochSwitchAtRound(round, parent)
 	if err != nil {
-		log.Error("[YourTurn] Failed while getting snapshot", "parentHash", parent.Hash(), "err", err)
+		log.Error("[YourTurn]", "Error", err)
 		return 0, -1, -1, false, err
 	}
-	masternodes := x.GetMasternodes(chain, parent)
-	if len(masternodes) == 0 {
-		return 0, -1, -1, false, errors.New("Masternodes not found")
-	}
-	pre := common.Address{}
-	// masternode[0] has chance to create block 1
-	preIndex := -1
-	if parent.Number.Uint64() != 0 {
-		pre, err = whoIsCreator(snap, parent)
-		if err != nil {
-			return 0, 0, 0, false, err
+	var masterNodes []common.Address
+	if isEpochSwitch {
+		if x.config.XDPoSV2Block.Cmp(parent.Number) == 0 {
+			// TODO: read v1 master nodes
+		} else {
+			// TODO: calc master nodes by smart contract - penalty
+			// TODO: related to snapshot
```

Comment:
Here is the TODOs that cannot be implemented easily

Reply:
Let's chat in zoom, but i suspect this piece of code shouldn't be here